### PR TITLE
Civ 3405

### PR DESCRIPTION
**Pull request by deepthidoppalapudihmcts**
- Replace existing pba call to service request
- created a new event at camunda level for payment service request and a new callback handler for payment servcie request
- Added new post mapping with service-request-reference to call credit account payment

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/CIV-3405

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/civil/service/PaymentsService.java [hunk 1]
```diff
@@ -58,20 +81,14 @@ private CreditAccountPaymentRequest buildRequest(CaseData caseData) {
...
             .accountNumber(generalAppPBADetails.getApplicantsPbaAccounts()
                     .getValue().getLabel())
             .amount(claimFee.getCalculatedAmount())
-            .caseReference(caseData.getLegacyCaseReference())
-            .ccdCaseNumber(caseData.getCcdCaseReference().toString())
             .customerReference(customerReference)
-            .description("Claim issue payment")
             .organisationName(organisationName)
-            .service(paymentsConfiguration.getService())
-            .siteId(paymentsConfiguration.getSiteId())
```
Comment (vasudevganesanhmcts) [inline src/main/java/uk/gov/hmcts/reform/civil/service/PaymentsService.java:73]: payments configuration site id is what we need to pass in the org id of the service request

File: src/main/java/uk/gov/hmcts/reform/civil/service/PaymentServiceRequest.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    private CasePaymentRequestDto casePaymentRequest;
+    @JsonProperty("case_reference")
+    private String caseReference;
+    @JsonProperty("ccd_case_number")
+    private String ccdCaseNumber;
+    @JsonProperty("fees")
+    private FeeDto[] fees;
+    @Builder.Default
+    @JsonProperty("hmcts_org_id")
+    private String hmctsOrgId = "ABA5";
```
Comment (vasudevganesanhmcts) [inline src/main/java/uk/gov/hmcts/reform/civil/service/PaymentServiceRequest.java:28]: hardcoded from private law not applicable to us

File: src/main/java/uk/gov/hmcts/reform/civil/handler/callback/camunda/payment/PaymentServiceRequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import static uk.gov.hmcts.reform.civil.callback.CaseEvent.MAKE_PAYMENT_SERVICE_REQ_GASPEC;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class PaymentServiceRequestHandler extends CallbackHandler {
+
+    private static final List<CaseEvent> EVENTS = Collections.singletonList(MAKE_PAYMENT_SERVICE_REQ_GASPEC);
+    private static final String ERROR_MESSAGE = "Technical error occurred";
+    private static final String TASK_ID = "GeneralApplicationMakePayment";
```
Comment (smitatalawadekar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/civil/handler/callback/camunda/payment/PaymentServiceRequestHandler.java:34]: This task id is bit confusing and not reflecting what the task is meant to do... would recommend renaming task id

File: src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java [hunk 4]
```diff
@@ -17,19 +17,19 @@
...
-    private final String pbaReference;
+    private final String serviceReqReference;
     private final Fee fee;
     private final PaymentDetails paymentDetails;
     private final LocalDateTime paymentSuccessfulDate;
 
     @JsonCreator
     GAPbaDetails(@JsonProperty("applicantsPbaAccounts") DynamicList applicantsPbaAccounts,
-                 @JsonProperty("pbaReference") String pbaReference,
+                 @JsonProperty("pbaReference") String serviceReqReference,
```
Comment (smitatalawadekar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java:27]: Is corresponding field changed in CCD? If not i would suggest to keep it in sync as it gets difficult to track field in pojo against field in ccd.... Also if the field in ccd is changed then this piece of code will not work as there will not be any json property pbaReference

File: src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java [hunk 5]
```diff
@@ -17,19 +17,19 @@
...
                  @JsonProperty("fee") Fee fee,
                  @JsonProperty("paymentDetails") PaymentDetails paymentDetails,
                  @JsonProperty("paymentSuccessfulDate") LocalDateTime paymentSuccessfulDate) {
         this.applicantsPbaAccounts = applicantsPbaAccounts;
-        this.pbaReference = pbaReference;
+        this.serviceReqReference = serviceReqReference;
```
Comment (smitatalawadekar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java:32]: Also, I am not sure if we should wipe out the PBA reference, that we accept from the user via screen, with service reference. Any specific reason for this change?

File: src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java (hunk 4 above)
Comment (deepthidoppalapudihmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java:27]: I have not changed in ccd you said that will not work as there will not be any json property with pbaReference

Comment (smitatalawadekar-hmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java:27]: pbaReference is captured on screen... so I am trying to understand why we are overriding it with service reference?

Comment (smitatalawadekar-hmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java:27]: also for traceability it is better to have same name in pojo and in CCD

File: src/main/java/uk/gov/hmcts/reform/civil/service/PaymentsService.java [hunk 6]
```diff
@@ -58,20 +90,15 @@ private CreditAccountPaymentRequest buildRequest(CaseData caseData) {
 
         String customerReference = ofNullable(generalAppPBADetails.getPaymentDetails())
             .map(PaymentDetails::getCustomerReference)
-            .orElse(generalAppPBADetails.getPbaReference());
```
Comment (vasudevganesanhmcts) [inline src/main/java/uk/gov/hmcts/reform/civil/service/PaymentsService.java:61]: needs to be correct to always use pba reference

File: src/main/java/uk/gov/hmcts/reform/civil/model/CasePaymentRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class CasePaymentRequestDto {
```
[the remark this answers is not in the document]
Comment (vasudevganesanhmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/model/CasePaymentRequestDto.java:15]: check on library

File: src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java (hunk 5 above)
Comment (deepthidoppalapudihmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java]: As discussed, Retained the value Pbarefrence and added a new field Service request Reference in GApbaDetails to store the service request response

File: src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java (hunk 4 above)
Comment (deepthidoppalapudihmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/model/genapplication/GAPbaDetails.java]: Yes, added a new field serviceReqReference in GAPbaDetails  in General application and in General ccd  and retained Pbareference as it was

File: src/main/java/uk/gov/hmcts/reform/civil/handler/callback/camunda/payment/PaymentServiceRequestHandler.java (hunk 3 above)
Comment (deepthidoppalapudihmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/handler/callback/camunda/payment/PaymentServiceRequestHandler.java]: Modified the name of Task_ID to GeneralApplicationPaymentServiceReq

File: src/main/java/uk/gov/hmcts/reform/civil/service/PaymentServiceRequest.java (hunk 2 above)
Comment (deepthidoppalapudihmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/service/PaymentServiceRequest.java]: Removed hardcoding and fetching value from paymentConfiguration class siteId

File: src/main/java/uk/gov/hmcts/reform/civil/service/PaymentsService.java (hunk 1 above)
Comment (deepthidoppalapudihmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/service/PaymentsService.java:73]: Done changed and committed

File: src/main/java/uk/gov/hmcts/reform/civil/model/CasePaymentRequestDto.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (deepthidoppalapudihmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/model/CasePaymentRequestDto.java]: Removed the class and its available on Payment client api lib

File: src/main/java/uk/gov/hmcts/reform/civil/service/PaymentsService.java (hunk 6 above)
Comment (deepthidoppalapudihmcts) [reply src/main/java/uk/gov/hmcts/reform/civil/service/PaymentsService.java:61]: Removed the mapping and mapping directly to the pbaRference

**Review by smitatalawadekar-hmcts — APPROVED** [review]

**Review by vasudevganesanhmcts — APPROVED** [review]

**Pull request merged**